NBA teams are increasingly trotting out lineups with five players who can play and guard nearly any position. traditional positions don't accurately explain what a players skillset truly is, they incorrectly oversimplify the skill sets of NBA players. Simply plugging players into one of five positions does not accurately define a player’s specific skill set. Moreover, the misclassification of a player’s position may lead teams to waste resources on developing draft picks that do not fit their systems.

In light of these changes, we need an effective way to designate positions in the NBA not based on basic physical traits such as height and weight, but in terms of function, such as shooting and defense. A framework for modern NBA positions is important towards our understanding for how players have evolved, and effective roster construction. 

### Import the Required Python Packages and Methods

In [92]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import csv

### Finding and Evaluating Historical NBA Player Data

I will use the Kaggle dataset "NBA Players stats since 1950", with stats for all players since 1950. The file Seasons_Stats.csv contains the statics of all players since 1950, with details of each column as below:


    Year - Season
    Player - name
    Pos - Position
    Age - Age
    Tm - Team
    G - Games
    GS - Games Started
    MP - Minutes Played
    PER - Player Efficiency Rating
    TS% - True Shooting %
    3PAr - 3-Point Attempt Rate
    FTr - Free Throw Rate
    ORB% - Offensive Rebound Percentage
    DRB% - Defensive Rebound Percentage
    TRB% - Total Rebound Percentage
    AST% - Assist Percentage
    STL% - Steal Percentage
    BLK% - Block Percentage
    TOV% - Turnover Percentage
    USG% - Usage Percentage
    blanl
    OWS - Offensive Win Shares
    DWS - Defensive Win Shares
    WS - Win Shares
    WS/48 - Win Shares Per 48 Minutes
    blank2
    OBPM - Offensive Box Plus/Minus
    DBPM - Defensive Box Plus/Minus
    BPM - Box Plus/Minus
    VORP - Value Over Replacement
    FG - Field Goals
    FGA - Field Goal Attempts
    FG% - Field Goal Percentage
    3P - 3-Point Field Goals
    3PA - 3-Point Field Goal Attempts
    3P% - 3-Point Field Goal Percentage
    2P - 2-Point Field Goals
    2PA - 2-Point Field Goal Attempts
    2P% - 2-Point Field Goal Percentage
    eFG% - Effective Field Goal Percentage
    FT - Free Throws
    FTA - Free Throw Attempts
    FT% - Free Throw Percentage
    ORB - Offensive Rebounds
    DRB - Defensive Rebounds
    TRB - Total Rebounds
    AST - Assists
    STL - Steals
    BLK - Blocks
    TOV - Turnovers
    PF - Personal Fouls
    PTS - Points

In [93]:
with open('data/Seasons_Stats.csv','rb') as csvfile:
    reader1 = csv.reader(csvfile)
df1 = pd.read_csv('data/Seasons_Stats.csv')

FileNotFoundError: [Errno 2] No such file or directory: 'data/Seasons_Stats.csv'

First, we remove duplicated or empty rows and drop a couple of blank columns.

In [68]:
df1_no_duplicates = df1.drop_duplicates()
df2 = df1_no_duplicates
del df2['blanl']
del df2['blank2']
df2.dropna(how='all')
df2.head(5)

Unnamed: 0    Year           Player  Pos   Age   Tm     G  GS  MP  PER  \
0           0  1950.0  Curly Armstrong  G-F  31.0  FTW  63.0 NaN NaN  NaN   
1           1  1950.0     Cliff Barker   SG  29.0  INO  49.0 NaN NaN  NaN   
2           2  1950.0    Leo Barnhorst   SF  25.0  CHS  67.0 NaN NaN  NaN   
3           3  1950.0       Ed Bartels    F  24.0  TOT  15.0 NaN NaN  NaN   
4           4  1950.0       Ed Bartels    F  24.0  DNN  13.0 NaN NaN  NaN   

     TS%  3PAr    FTr  ORB%  DRB%  TRB%  AST%  STL%  BLK%  TOV%  USG%  OWS  \
0  0.368   NaN  0.467   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN -0.1   
1  0.435   NaN  0.387   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  1.6   
2  0.394   NaN  0.259   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  0.9   
3  0.312   NaN  0.395   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN -0.5   
4  0.308   NaN  0.378   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN -0.5   

   DWS   WS  WS/48  OBPM  DBPM  BPM  VORP     FG    FGA    FG%  3P  3PA  3P%  \
0  3.6  3.5    NaN   NaN   NaN  NaN   NaN  144.0  516.0  0.279 NaN  NaN  NaN   
1  0.6  2.2    NaN   NaN   NaN  NaN   NaN  102.0  274.0  0.372 NaN  NaN  NaN   
2  2.8  3.6    NaN   NaN   NaN  NaN   NaN  174.0  499.0  0.349 NaN  NaN  NaN   
3 -0.1 -0.6    NaN   NaN   NaN  NaN   NaN   22.0   86.0  0.256 NaN  NaN  NaN   
4 -0.1 -0.6    NaN   NaN   NaN  NaN   NaN   21.0   82.0  0.256 NaN  NaN  NaN   

      2P    2PA    2P%   eFG%     FT    FTA    FT%  ORB  DRB  TRB    AST  STL  \
0  144.0  516.0  0.279  0.279  170.0  241.0  0.705  NaN  NaN  NaN  176.0  NaN   
1  102.0  274.0  0.372  0.372   75.0  106.0  0.708  NaN  NaN  NaN  109.0  NaN   
2  174.0  499.0  0.349  0.349   90.0  129.0  0.698  NaN  NaN  NaN  140.0  NaN   
3   22.0   86.0  0.256  0.256   19.0   34.0  0.559  NaN  NaN  NaN   20.0  NaN   
4   21.0   82.0  0.256  0.256   17.0   31.0  0.548  NaN  NaN  NaN   20.0  NaN   

   BLK  TOV     PF    PTS  
0  NaN  NaN  217.0  458.0  
1  NaN  NaN   99.0  279.0  
2  NaN  NaN  192.0  438.0  
3  NaN  NaN   29.0   63.0  
4  NaN  NaN   27.0   59.0

A second file, players.csv, contains static information for each player, as height, weight, etc.

In [69]:
players = pd.read_csv('data/Players.csv', index_col=0)
players.head(5)

Player  height  weight                          collage    born  \
0  Curly Armstrong   180.0    77.0               Indiana University  1918.0   
1     Cliff Barker   188.0    83.0           University of Kentucky  1921.0   
2    Leo Barnhorst   193.0    86.0         University of Notre Dame  1924.0   
3       Ed Bartels   196.0    88.0  North Carolina State University  1925.0   
4      Ralph Beard   178.0    79.0           University of Kentucky  1927.0   

    birth_city birth_state  
0          NaN         NaN  
1     Yorktown     Indiana  
2          NaN         NaN  
3          NaN         NaN  
4  Hardinsburg    Kentucky

I then selected these three data categories:
    
    Player - Player's full name (first and last)
    height - Height in cm
    weight - Weight in kg

1.The players have unique names (checked at the beginning), we can merge the two dataframes using the 'Player' column.

In [70]:
data = pd.merge(df2, players[['Player', 'height', 'weight']], left_on='Player', right_on='Player', right_index=False,
      how='left', sort=False)
#.fillna(value=0)

2.Using rate statistics (i.e. points per game) or cumulative statistics (i.e. total points) can be misleading when it comes to analysis because these statistics tend to inflate players with lengthier careers. To deal with outliers, I instituted a minimum threshold of 40 games played. Also I keep only players with more than 400 minutes for each season (with a 82 games regular season, thats around 5 minutes per game. Players with less than that will be only anecdotical, and will distort the analysis).

In [71]:
data = data[~(data['Pos']==0) & (data['MP'] > 400) & (data['G'] > 40)]
data.reset_index(inplace=True, drop=True)

3.Replace the * sign in some of the names.

In [72]:
data['Player'] = data['Player'].str.replace('*','')

4.For the stats that represent total values (others, as TS%, represent percentages), we will take the values per 36 minutes. The reason is to judge every player according to his characteristics, not the time he was on the floor.

In [73]:
totals = ['PER', 'OWS', 'DWS', 'WS', 'OBPM', 'DBPM', 'BPM', 'VORP', 'FG', 'FGA', '3P', '3PA', '2P', '2PA', 'FT', 'FTA',
         'ORB', 'DRB', 'TRB', 'AST', 'STL', 'BLK', 'TOV', 'PF', 'PTS']

for col in totals:
    data[col] = 36 * data[col] / data['MP']

In [75]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15383 entries, 0 to 15382
Data columns (total 53 columns):
Unnamed: 0    15383 non-null int64
Year          15383 non-null float64
Player        15383 non-null object
Pos           15383 non-null object
Age           15381 non-null float64
Tm            15383 non-null object
G             15383 non-null float64
GS            11192 non-null float64
MP            15383 non-null float64
PER           15365 non-null float64
TS%           15383 non-null float64
3PAr          11668 non-null float64
FTr           15383 non-null float64
ORB%          12997 non-null float64
DRB%          12997 non-null float64
TRB%          13543 non-null float64
AST%          14260 non-null float64
STL%          12997 non-null float64
BLK%          12997 non-null float64
TOV%          12165 non-null float64
USG%          12165 non-null float64
OWS           15365 non-null float64
DWS           15365 non-null float64
WS            15365 non-null float64
WS/48   

from the data information above, the ideal number of data rows we want to work with is 15383, most amount of data is missing from 

    3P  (3-Point Field Goals)                  11668 non-null float64
    3PA (3-Point Field Goal Attempts)          11668 non-null float64
    3P% (3-Point Field Goal Percentage)        10597 non-null float64
    3PAr(3-Point Attempt Rate)                 11668 non-null float64
    
It's becasue 3-point line was not introduced to NBA until 1979, but for the sake of classifying modern players, I have to assume all players before 1979 have not attempted 3-pointers, so I will fill the missing values of these three columns with 0. 

In addition:

    GS  (Games Started)                        11192 non-null float64

Since number of Games started for each player does not effect their performance measuring, I will delete this column.

In [76]:
threePointsCol = ['3P','3PA','3P%','3PAr']
for feature in threePointsCol:
    data[feature] = data[feature].fillna(0)
del data['GS']
data[threePointsCol].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15383 entries, 0 to 15382
Data columns (total 4 columns):
3P      15383 non-null float64
3PA     15383 non-null float64
3P%     15383 non-null float64
3PAr    15383 non-null float64
dtypes: float64(4)
memory usage: 480.8 KB


The next groups of columns that have substantical amount of missing data is from:
    
    ORB           12997 non-null float64
    DRB           12997 non-null float64
    STL           12997 non-null float64
    BLK           12997 non-null float64
    TOV           12165 non-null float64
    
And their associate percentages(plus USG%, relevant to TOV):
    
    ORB%          12997 non-null float64
    DRB%          12997 non-null float64
    STL%          12997 non-null float64
    BLK%          12997 non-null float64
    TOV%          12165 non-null float64
    USG%          12165 non-null float64
    
Since the above data has minimul 12165 or 12997 rows, the commonality is due to the fact ORB, DRB, STL, BLK and their associate percentages are not included in this dataset before 1974, and TOV data are not present until 1978. These missing values cannot be filled with assumption, but at the same time these columns are too important to be removed from the construction of player classification, therefore all the data before 1978 have to be discarded. 

In [89]:
data = data[~(data['Year'] < 1978)]

5.Cast these columns to be of type 'int'.

In [90]:
integerCol = ['Year','Age', 'G']
for feature in integerCol:
    data[feature] = data[feature].astype(dtype ='int')

/Users/yidingweng/anaconda/lib/python3.6/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [91]:
pd.options.display.max_columns = None
data.sample(5).transpose()

11821            11106          12350          7464   \
Unnamed: 0       18590            17410          19512          11330   
Year              2007             2005           2009           1993   
Player      Chris Paul  Antonio McDyess  Jordan Farmar  Herb Williams   
Pos                 PG               PF             PG              C   
Age                 21               30             22             34   
Tm                 NOK              DET            LAL            NYK   
G                   64               77             65             55   
MP                2353             1797           1192            571   
PER           0.336592         0.344574       0.298993       0.636778   
TS%              0.537            0.542          0.466          0.429   
3PAr             0.164            0.002          0.354              0   
FTr               0.41            0.321          0.187           0.12   
ORB%               2.6             11.6            2.1            8.8   
DRB%              11.3             19.7            8.9           19.3   
TRB%               6.9             15.7            5.5           14.2   
AST%              41.3              7.3           18.6            4.9   
STL%               2.7              1.4            2.4            1.9   
BLK%               0.1              2.2            0.6            3.3   
TOV%              13.5               12           16.3           10.7   
USG%              23.1             20.3           19.6           15.6   
OWS          0.0948576        0.0500835     -0.0211409     -0.0189142   
DWS           0.039779        0.0661102       0.045302      0.0882662   
WS            0.134637          0.11419      0.0241611       0.069352   
WS/48            0.179            0.154          0.034          0.095   
OBPM         0.0688483       -0.0160267     -0.0724832      -0.321541   
DBPM        0.00152996        0.0200334     -0.0151007       0.208056   
BPM          0.0703782       0.00400668     -0.0875839      -0.113485   
VORP         0.0596685        0.0200334     -0.0090604              0   
FG             5.82915          6.15025        4.86242         4.5394   
FGA             13.326            11.98         12.443        11.0333   
FG%              0.437            0.513          0.391          0.411   
3P            0.764981                0        1.47987              0   
3PA            2.18785        0.0200334         4.4094              0   
3P%               0.35                0          0.336              0   
2P             5.06417          6.15025        3.38255         4.5394   
2PA            11.1381          11.9599        8.03356        11.0333   
2P%              0.455            0.514          0.421          0.411   
eFG%             0.466            0.513           0.45          0.411   
FT             4.46749          2.52421        1.35906       0.882662   
FTA            5.46196          3.84641         2.3255        1.32399   
FT%              0.818            0.656          0.584          0.667   
ORB           0.826179          3.54591        0.66443        2.77408   
DRB            3.45771          6.11018        2.86913        6.43082   
TRB            4.28389          9.65609        3.53356         9.2049   
AST            8.70548          1.42237        4.68121         1.1979   
STL            1.80535         0.921536        1.72148        1.32399   
BLK          0.0458989          1.04174       0.302013        1.76532   
TOV            2.46324          1.86311        2.62752        1.38704   
PF             2.35614          4.08681        3.17114        4.91769   
PTS            16.8908          14.8247        12.5638        9.96147   
height             183              206            188            208   
weight              79               99             81            109   

                      8392   
Unnamed: 0            12807  
Year                   1996  
Player      Jayson Williams  
Pos                      PF

In [80]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12997 entries, 2386 to 15382
Data columns (total 52 columns):
Unnamed: 0    12997 non-null int64
Year          12997 non-null float64
Player        12997 non-null object
Pos           12997 non-null object
Age           12997 non-null float64
Tm            12997 non-null object
G             12997 non-null float64
MP            12997 non-null float64
PER           12997 non-null float64
TS%           12997 non-null float64
3PAr          12997 non-null float64
FTr           12997 non-null float64
ORB%          12997 non-null float64
DRB%          12997 non-null float64
TRB%          12997 non-null float64
AST%          12997 non-null float64
STL%          12997 non-null float64
BLK%          12997 non-null float64
TOV%          12165 non-null float64
USG%          12165 non-null float64
OWS           12997 non-null float64
DWS           12997 non-null float64
WS            12997 non-null float64
WS/48         12997 non-null float64
OBPM 